
#  This file is used to configure the paramters used to control the servos.
#  This is an example only. To use the controller define a new controller_config.yaml
#  in your ros package along with a new launch file and use those. see the readme for
#  more info.

# # --------------------------------------------------------------------------- # #   

# GLOBAL OPERATION PARAMETERS

publish_rate: 200                  # desired rate for joint state updates. actual rate may be less depending on number
                                    # of dynamixels on each port

control_mode: Position                # control mode, either 'Position', 'Velocity', or 'Torque'

dynamic_mode_switching: true       # enables dynamixe switching between torque and position control when in position control mode
                                    # (CAUTION: EXPERIMENTAL)

disable_torque_on_shutdown: true    # with this enabled the motors will switch off when the controller closes

echo_joint_commands: false           # debug flag to echo write commands sent to the controller, useful for measuring write rates

use_torque_as_effort: true         # when sending and receiving joint states, report effort as torque or raw % of max

mx_effort_use_current: true        # Use the current register instead of the load register on the mx series (IO performance hit)

diagnostics_rate: 0                 # rate to publish diagnostic information

# The below values are used as global defaults and are applied for each servo unless overridden in the entry for 
# the servo below

global_joint_speed: 5.0             # maximum joint speed (rad/s) (in position or velocity control)

global_torque_limit: 1.0            # maximum motor torque for all modes, given as a fraction of rated max (0-1)

global_p_gain: -1.0                 # proportional gain value (values > 0 are set, -1.0 indicates to leave on default)
global_i_gain: -1.0                 # integral gain value (values > 0 are set, -1.0 indicates to leave on default)
global_d_gain: -1.0                 # derivative gain value (values > 0 are set, -1.0 indicates to leave on default)


# # --------------------------------------------------------------------------- # #   

# PORT AND SERVO CONFIGURATIONS
ports:

  # PORT LIST

  - name: Port_1                    # name for this port in config
    device: /dev/ttyUSB1            # serial device this port will communicate on
    baudrate: 1000000               # baudrate in use (see dynamixel_driver.h for valid values)
    protocol: "1.0"                 # motor protocol in use, must be "1.0", "2.0", or "PRO"
    group_comms_enabled: true
    servos:
        
        # SERVO LIST FOR THIS PORT

        - id: 1                     # (ID set in servo eeprom, must be unique on this port)
          joint_name: joint_1       # (MUST BE UNIQUE ACROSS ALL PORTS)
          # 
          # The three values below are mandatory, they define the orientation and zeroing of the dynamixel:
          # 
          init: 2048                # initial (0 rad) servo position (in raw encoder count)
          min: 0                    # minimum servo position (in raw encoder count) 
          max: 4095                 # maximum servo position, Note when MIN > MAX ROTATION IS REVERSED
          # 
          # The below arguments are all optional and override the global values:
          # 
          joint_speed: 5.0          # maximum joint speed (rad/s) (in position or velocity control)
          torque_limit: 1.0         # maximum motor torque for all modes, given as a fraction of rated max (0-1)
          p_gain: -1.0              # proportional gain value (values > 0 are set, -1.0 indicates to leave on default)
          i_gain: -1.0              # integral gain value (values > 0 are set, -1.0 indicates to leave on default)
          d_gain: -1.0              # derivative gain value (values > 0 are set, -1.0 indicates to leave on default)


        - id: 2
          joint_name: joint_2
          init: 2048
          min: 0
          max: 4095
          # 
          # This servo doesn't have any optional values defined, the global defaults will be used
          #

  # ------------------------------------------------------------------------- #   

  # - name: Port_2                    # more ports can be defined in the same manner as above, each port can even have a 
  #   device: /dev/ttyUSB1            # different baudrate and protcol
  #   baudrate: 2000000
  #   protocol: "2.0"
  #   servos:

  #       - id: 1                     # id 1 only needs to be unique for each port and so can be reused here
  #         joint_name: joint_3
  #         init: 1000
  #         min: 1000
  #         max: 3000
  #         joint_speed: 1.0
  #         torque_limit: 0.5
  #         p_gain: 1.0
  #         i_gain: 0.0
  #         d_gain: 0.0

  #       - id: 2
  #         joint_name: joint_4
  #         init: 2048
  #         min: 0
  #         max: 4095
  #         joint_speed: 5.0
  #         torque_limit: 1.0
  #         p_gain: -1.0
  #         i_gain: -1.0
  #         d_gain: -1.0

# # --------------------------------------------------------------------------- # #   